### DISCUSSION TITLE
[ObservableProperty] cannot be used when inheriting from BaseViewModel with INotifyPropertyChanged implemented

### DISCUSSION THREAD

**Discussion by jakubjenis** [CommunityToolkit/dotnet#336] [Ideas]
When trying to utilize the awesome [ObservableProperty] code generator, I have hit a roadblock caused by my ViewModels having a base class from FreshMvvm library - FreshBasePageModel.

``` c#
public partial class SampleViewModelBrokenWithInheritance : FreshMvvm.Maui.FreshBasePageModel
{
  [ObservableProperty]
  int _sampleProperty;
}
```

This does not compile with a following message: 

> MVVMTK0019	The field MauiFreshMvvmDemo.SampleViewModelBrokenWithInheritance._sampleProperty cannot be used to generate an observable property, as its containing type doesn't inherit from ObservableObject, nor does it use [ObservableObject] or [INotifyPropertyChanged]. 

I cannot inherit from ObservableObject as I already have a base class. If I try to add the [ObservableObject] or [INotifyPropertyChanged] attributes, I get this: 

> MVVMTK0001	Cannot apply [INotifyPropertyChanged] to type MauiFreshMvvmDemo.SampleViewModelBrokenWithInheritance, as it already declares the INotifyPropertyChanged interface

Shouldn't the code generator leverage my base class already implementing INotifyPropertyChanged interface? When looking at the generated code, I kind of have the idea that [ObservableProperty] is calling into some generated OnPropertyChanged method that is added by [INofifyPropertyChanged] code generator.

``` c#
[global::System.CodeDom.Compiler.GeneratedCode("CommunityToolkit.Mvvm.SourceGenerators.ObservablePropertyGenerator", "8.0.0.0")]
[global::System.Diagnostics.CodeAnalysis.ExcludeFromCodeCoverage]
public int SampleProperty
{
    get => _sampleProperty;
    set
    {
        if (!global::System.Collections.Generic.EqualityComparer<int>.Default.Equals(_sampleProperty, value))
        {
            OnSamplePropertyChanging(value);
            _sampleProperty = value;
            OnSamplePropertyChanged(value);
            OnPropertyChanged(global::CommunityToolkit.Mvvm.ComponentModel.__Internals.__KnownINotifyPropertyChangedArgs.SampleProperty);
        }
    }
}
```

Wouldn't it be better if instead of calling the OnPropertyChanged method, it was just invoking PropertyChanged event directly (it is doing this anyway, just wrapped in the OnPropertyChanged method). This would then no longer require [INotifyPropertyChanged] attribute being applied to my viewModel class and [ObservableProperty] could be applied even when inheriting from FreshBasePageModel.

The alternative is for the FreshMvvm library (or any other MVVM framework you are using) to apply the [INotifyPropertyChanged] attribute in its base class, but with my proposed approach this is no longer required. Isn't the whole idea of INotifyPropertyChanged to provide this loose coupling?

**Comment by chrispeabody** [comment]
I am hitting this same issue when making custom controls in .NET maui while utilizing the `ContentView` class. The `BindableObject` class, which is a base class of `ContentView`, implements `INotifyPropertyChanged`. If I try to use the `ObservableObject` or `INotifyPropertyChanged` attributes, it gives me the `MVVMTK0002` error, which appears to have identical text to `MVVMTK0001` If I do not use those attributes, I get `MVVMTK0019`, as above. 

Notably, the `INofityPropertyChanged` interface that `BindableObject` implements appears to be from `System.ComponentModel` and not the MVVM library. I'm not sure how this would be remedied...

**Comment by PRNDA** [comment]
I am hitting this same issue too.

**Reply by chrispeabody** [reply]
A year later, I still have this issue in projects. I'm wondering if there is a solution I just haven't found yet?

**Comment by AzzaouiAlae** [comment]
i have the same problem but i just use MAUI binding 

```
public static readonly BindableProperty CategoryNameProperty =
    BindableProperty.Create(nameof(CategoryName), typeof(string), typeof(TransactionView), null);
public string CategoryName
{
    get => (string)GetValue(CategoryNameProperty);
    set => SetValue(CategoryNameProperty, value);
}  
```

**Comment by hawkerm** [comment]
There's an issue tracking this here: https://github.com/CommunityToolkit/dotnet/issues/620
